{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "XSQPhLCv_wVq"
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/MiqG/ColabSplice/blob/main/notebooks/Borzoi.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "-6T7ynmAAjjx"
   },
   "source": [
    "# **Borzoi**\n",
    "\n",
    "## Model information\n",
    "- original repository: https://github.com/calico/borzoi\n",
    "- original paper: https://doi.org/10.1038/s41588-024-02053-6\n",
    "- pytorch version:\n",
    "  - repository: https://github.com/johahi/borzoi-pytorch\n",
    "  - paper: https://doi.org/10.1101/2024.12.18.629121\n",
    "\n",
    "## Tips and Instructions\n",
    "- click the little â–¶ play icon to the left of each cell below. Or Runtime > Run all to give it a try. This will produce a zip file called `test_SMN2.zip`, containing:\n",
    "  - splice site strength predictions:\n",
    "    - table: `test_SMN2-scores.tsv`\n",
    "    - plot: `test_SMN2-coverage-line.pdf`\n",
    "  - bigwig files that can be visualized through [IGV](https://igv.org/):\n",
    "    - `test_SMN2-RNA:adipose_tissue.bw`\n",
    "    - `test_SMN2-RNA:adrenal_gland.bw`\n",
    "    - `test_SMN2-RNA:bladder.bw`\n",
    "    - `test_SMN2-RNA:blood.bw`\n",
    "    - `test_SMN2-RNA:blood_vessel.bw`\n",
    "    - `test_SMN2-RNA:brain.bw`\n",
    "    - `test_SMN2-RNA:breast.bw`\n",
    "    - `test_SMN2-RNA:cervix_uteri.bw`\n",
    "    - `test_SMN2-RNA:colon.bw`\n",
    "    - `test_SMN2-RNA:esophagus.bw`\n",
    "    - `test_SMN2-RNA:fallopian_tube.bw`\n",
    "    - `test_SMN2-RNA:heart.bw`\n",
    "    - `test_SMN2-RNA:kidney.bw`\n",
    "    - `test_SMN2-RNA:liver.bw`\n",
    "    - `test_SMN2-RNA:lung.bw`\n",
    "    - `test_SMN2-RNA:muscle.bw`\n",
    "    - `test_SMN2-RNA:nerve.bw`\n",
    "    - `test_SMN2-RNA:ovary.bw`\n",
    "    - `test_SMN2-RNA:pancreas.bw`\n",
    "    - `test_SMN2-RNA:pituitary.bw`\n",
    "    - `test_SMN2-RNA:prostate.bw`\n",
    "    - `test_SMN2-RNA:salivary_gland.bw`\n",
    "    - `test_SMN2-RNA:skin.bw`\n",
    "    - `test_SMN2-RNA:small_intestine.bw`\n",
    "    - `test_SMN2-RNA:spleen.bw`\n",
    "    - `test_SMN2-RNA:stomach.bw`\n",
    "    - `test_SMN2-RNA:testis.bw`\n",
    "    - `test_SMN2-RNA:thyroid.bw`\n",
    "    - `test_SMN2-RNA:uterus.bw`\n",
    "    - `test_SMN2-RNA:vagina.bw`\n",
    "     \n",
    "- The notebooks runs perfectly without GPU.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "7KnUYg24_Tkp"
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "#@title ## Install packages (~1min 30s)\n",
    "import os, time\n",
    "if not os.path.isfile(\"finished_install\"):\n",
    "  # genomic file handling\n",
    "  os.system(\"pip install -q pyfaidx==0.8.1.1 pyranges pyBigWig pyrle biopython\")\n",
    "\n",
    "  # install Borzoi\n",
    "  os.system(\"pip install -q borzoi-pytorch\")\n",
    "\n",
    "  # install visualization\n",
    "  os.system(\"pip install -q igv-notebook\")\n",
    "\n",
    "  os.system(\"touch finished_install\")\n",
    "\n",
    "# imports\n",
    "import pyfaidx\n",
    "import pyranges as pr\n",
    "import hashlib, re, os\n",
    "import torch\n",
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import igv_notebook\n",
    "import shutil\n",
    "from google.colab import files\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from borzoi_pytorch import Borzoi\n",
    "from borzoi_pytorch.config_borzoi import BorzoiConfig\n",
    "from borzoi_pytorch.pytorch_borzoi_helpers import predict_tracks\n",
    "import random\n",
    "\n",
    "# notebook-wide variables\n",
    "MAX_SEQ_LEN = 524288\n",
    "\n",
    "targets = pd.read_table(\"https://github.com/johahi/borzoi-pytorch/raw/refs/heads/main/borzoi_pytorch/precomputed/targets.txt\")\n",
    "targets = targets.loc[targets[\"identifier\"].str.contains(\"GTEX\")]\n",
    "TISSUES = targets\n",
    "\n",
    "GENOME_FASTA_URLS = {\n",
    "    \"homo_sapiens (hg38-GencodeV44)\": \"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_44/GRCh38.primary_assembly.genome.fa.gz\"\n",
    "}\n",
    "\n",
    "# functions\n",
    "def install_genome(species, genome_fasta_urls, fasta_file):\n",
    "  if not os.path.isfile(\"finished_install_genome\"):\n",
    "    print(f\"Downloading genome for {species}...\")\n",
    "    os.system(f\"wget {genome_fasta_urls[species]} -O {fasta_file}\")\n",
    "\n",
    "    print(\"Uncompressing genome fasta...\")\n",
    "    os.system(f\"gunzip {fasta_file}\")\n",
    "\n",
    "    os.system(\"touch finished_install_genome\")\n",
    "\n",
    "def get_hash(x):\n",
    "  return hashlib.sha1(x.encode()).hexdigest()\n",
    "\n",
    "def save_fasta(sequence, filename, header=\"sequence\"):\n",
    "  record = SeqRecord(Seq(sequence), id=header, description=\"\")\n",
    "  with open(filename, \"w\") as f:\n",
    "    SeqIO.write(record, f, \"fasta\")\n",
    "\n",
    "def reverse_complement(seq, padval_seq=\"N\"):\n",
    "  complement = str.maketrans(padval_seq+\"ACGTacgt\", padval_seq+\"TGCAtgca\")\n",
    "  return seq.translate(complement)[::-1]\n",
    "\n",
    "def load_models():\n",
    "  cfg = BorzoiConfig.from_pretrained('johahi/borzoi-replicate-0')\n",
    "  cfg.return_center_bins_only = False\n",
    "  models = Borzoi.from_pretrained('johahi/borzoi-replicate-0', config=cfg)\n",
    "  return [models]\n",
    "\n",
    "def dna_1hot(\n",
    "    seq: str, seq_len: int = None, n_uniform: bool = False, n_sample: bool = False\n",
    "):\n",
    "  \"\"\"Convert a DNA sequence to a 1-hot encoding.\n",
    "\n",
    "  Args:\n",
    "    seq (str): DNA sequence.\n",
    "    seq_len (int): length to extend/trim sequences to.\n",
    "    n_uniform (bool): represent N's as 0.25, forcing float16,\n",
    "    n_sample (bool):  sample ACGT for N\n",
    "\n",
    "  Returns:\n",
    "    seq_code (np.array): 1-hot encoding of DNA sequence.\n",
    "  \"\"\"\n",
    "  if seq_len is None:\n",
    "    seq_len = len(seq)\n",
    "    seq_start = 0\n",
    "  else:\n",
    "    if seq_len <= len(seq):\n",
    "      # trim the sequence\n",
    "      seq_trim = (len(seq) - seq_len) // 2\n",
    "      seq = seq[seq_trim : seq_trim + seq_len]\n",
    "      seq_start = 0\n",
    "    else:\n",
    "      seq_start = (seq_len - len(seq)) // 2\n",
    "\n",
    "  seq = seq.upper()\n",
    "\n",
    "  # map nt's to a matrix len(seq)x4 of 0's and 1's.\n",
    "  if n_uniform:\n",
    "    seq_code = np.zeros((seq_len, 4), dtype=\"float16\")\n",
    "  else:\n",
    "    seq_code = np.zeros((seq_len, 4), dtype=\"bool\")\n",
    "\n",
    "  for i in range(seq_len):\n",
    "    if i >= seq_start and i - seq_start < len(seq):\n",
    "      nt = seq[i - seq_start]\n",
    "      if nt == \"A\":\n",
    "        seq_code[i, 0] = 1\n",
    "      elif nt == \"C\":\n",
    "        seq_code[i, 1] = 1\n",
    "      elif nt == \"G\":\n",
    "        seq_code[i, 2] = 1\n",
    "      elif nt == \"T\":\n",
    "        seq_code[i, 3] = 1\n",
    "      else:\n",
    "        if n_uniform:\n",
    "          seq_code[i, :] = 0.25\n",
    "        elif n_sample:\n",
    "          ni = random.randint(0, 3)\n",
    "          seq_code[i, ni] = 1\n",
    "\n",
    "  return seq_code\n",
    "\n",
    "def one_hot_encode(seq):\n",
    "  seq = dna_1hot(seq)\n",
    "  return seq\n",
    "\n",
    "def get_margins(seq_len):\n",
    "  margin = (MAX_SEQ_LEN - (seq_len)) / 2\n",
    "  if margin % 1 == 0:\n",
    "    margin_upstream = margin\n",
    "    margin_downstream = margin\n",
    "  else:\n",
    "    margin_upstream = int(margin)\n",
    "    margin_downstream = int(margin)+1\n",
    "\n",
    "  assert MAX_SEQ_LEN == margin_upstream + margin_downstream + seq_len\n",
    "\n",
    "  return int(margin_upstream), int(margin_downstream)\n",
    "\n",
    "def predict(sequence, strand, models):\n",
    "\n",
    "  print(\"Computing prediction across different tissues...\")\n",
    "\n",
    "  if strand == '-':\n",
    "    sequence = reverse_complement(sequence)\n",
    "\n",
    "  sequence = one_hot_encode(sequence).T\n",
    "  sequence = torch.from_numpy(sequence).float().unsqueeze(0)\n",
    "  if torch.cuda.is_available():\n",
    "    sequence = sequence.cuda()\n",
    "\n",
    "  scores = []\n",
    "  for fold_idx, model in enumerate(models):\n",
    "    model.eval()\n",
    "    with torch.no_grad():\n",
    "      # predict\n",
    "      output = model(sequence).cpu().detach()\n",
    "      output = output.numpy(force=True) # 1 x 7611 features x 6144 seq_bins\n",
    "      output = output[0,targets.index,:] # subset features of interest\n",
    "      scores.append(output)\n",
    "\n",
    "  if strand == '-':\n",
    "    # reverse the bins\n",
    "    scores = [o[:,::-1] for o in scores]\n",
    "\n",
    "  # prepare outputs\n",
    "  scores = [pd.DataFrame(score.T, columns=TISSUES[\"identifier\"]) for score in scores]\n",
    "  for idx in range(len(scores)):\n",
    "    scores[idx] = pd.merge(\n",
    "        scores[idx].T, TISSUES[[\"identifier\",\"description\"]],\n",
    "        left_index=True, right_on=\"identifier\", how=\"left\"\n",
    "    ).drop(\n",
    "        \"identifier\",\n",
    "        axis=1\n",
    "    ).groupby(\n",
    "        \"description\"\n",
    "    ).mean().T\n",
    "\n",
    "    scores[idx] = scores[idx].loc[scores[idx].index.repeat(32)].reset_index(drop=True)\n",
    "\n",
    "  return scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "ir1Rl6S2Cfe0"
   },
   "outputs": [],
   "source": [
    "#@title ##run **Borzoi**\n",
    "%%time\n",
    "\n",
    "# user inputs\n",
    "# @markdown Run settings\n",
    "jobname = \"test_SMN2\" #@param {type:\"string\"}\n",
    "jobname = re.sub(r'\\W+', '', jobname)[:50]\n",
    "\n",
    "predict_from = \"genomic_coordinates\" #@param [\"custom_sequence\",\"genomic_coordinates\"]\n",
    "\n",
    "download_outputs = True #@param {type:\"boolean\"}\n",
    "igv_view = True #@param {type:\"boolean\"}\n",
    "\n",
    "# @markdown ---\n",
    "\n",
    "if predict_from == \"genomic_coordinates\":\n",
    "  # @markdown Genomic coordinates (NOTE: requires installing genome at first run (~3min)):\n",
    "  # coordinates\n",
    "  chrom = \"chr5\" #@param {type:\"string\"}\n",
    "  start = 70049669 #@param {type:\"integer\"}\n",
    "  end = 70077595 #@param {type:\"integer\"}\n",
    "  strand = \"+\" #@param [\"-\",\"+\"]\n",
    "  margin_upstream, margin_downstream = get_margins(end - start)\n",
    "\n",
    "  # genome\n",
    "  species = \"homo_sapiens (hg38-GencodeV44)\" #@param [\"homo_sapiens (hg38-GencodeV44)\"]\n",
    "  fasta_file = os.path.basename(GENOME_FASTA_URLS[species])\n",
    "  install_genome(species, GENOME_FASTA_URLS, fasta_file)\n",
    "  fasta_file = fasta_file.replace(\".gz\",\"\")\n",
    "\n",
    "  print(\"Loading genome data...\")\n",
    "  genome = pyfaidx.Fasta(fasta_file)\n",
    "  sequence = genome[chrom][(start-margin_upstream):(end+margin_downstream)].seq\n",
    "\n",
    "elif predict_from == \"custom_sequence\":\n",
    "  # @markdown ---\n",
    "  # @markdown Custom pre-mRNA sequence (try copy-pasting [FAS sequence](https://www.ncbi.nlm.nih.gov/nuccore/NC_000010.11?report=fasta&from=88964050&to=89017059)):\n",
    "  sequence = \"\" # @param {\"type\":\"string\",\"placeholder\":\"Paste nucleotide sequence here\"}\n",
    "  sequence = sequence.replace(\" \",\"\")\n",
    "  chrom = f\"{jobname}\"\n",
    "  strand = \"+\"\n",
    "  margin_upstream, margin_downstream = get_margins(len(sequence))\n",
    "  sequence = margin_downstream*\"N\" + sequence + margin_upstream*\"N\"\n",
    "\n",
    "  # save custom sequence as fasta\n",
    "  fasta_file = f\"{jobname}.fa\"\n",
    "  save_fasta(sequence[512:-512], fasta_file, header=f\"{jobname}\")\n",
    "  genome = pyfaidx.Fasta(fasta_file)\n",
    "  start = 1\n",
    "  end = len(sequence) - margin_upstream - margin_downstream\n",
    "\n",
    "sequence = sequence.upper().replace(\"U\",\"T\")\n",
    "assert set(sequence).issubset({\"A\", \"C\", \"T\", \"G\", \"N\"})\n",
    "\n",
    "ID = jobname+\"_\"+get_hash(sequence)[:5]\n",
    "\n",
    "# load model\n",
    "models = load_models()\n",
    "if torch.cuda.is_available():\n",
    "  [model.cuda() for model in models]\n",
    "\n",
    "# make prediction\n",
    "print(f\"Using {predict_from} to make prediction.\")\n",
    "scores = predict(sequence, strand, models)\n",
    "\n",
    "# add more info\n",
    "positions = np.arange(start-margin_upstream+512, end+margin_downstream-512)\n",
    "if strand==\"-\":\n",
    "  sequence = reverse_complement(sequence)[::-1]\n",
    "\n",
    "for idx in range(len(scores)):\n",
    "  scores[idx][\"chrom\"] = chrom\n",
    "  scores[idx][\"position\"] = positions\n",
    "  scores[idx][\"sequence\"] = list(sequence[512:-512])\n",
    "\n",
    "# save\n",
    "scores = scores[0]\n",
    "filename_scores = f\"{jobname}-scores.tsv\"\n",
    "scores.to_csv(filename_scores, index=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "OxSVXKsmLGUQ"
   },
   "outputs": [],
   "source": [
    "#@title PDF plot\n",
    "\n",
    "X = scores.melt(id_vars=[\"chrom\",\"position\",\"sequence\"], var_name=\"tissue\")\n",
    "\n",
    "g = sns.relplot(data=X, x=\"position\", y=\"value\", hue=\"tissue\", row=\"tissue\", kind=\"line\", height=1.5, aspect=10)\n",
    "g.set_xlabels(\"\")\n",
    "g.set_ylabels(\"\", clear_inner=False)\n",
    "g.figure.text(0.5, 0.02, \"Nucleotide Position\", ha=\"center\", fontsize=10)\n",
    "g.figure.text(0.0, 0.5, \"Predicted Splice Site Strength\", va=\"center\", rotation=\"vertical\", fontsize=10)\n",
    "\n",
    "filename_plt = f\"{jobname}-coverage-line.pdf\"\n",
    "g.figure.savefig(filename_plt, format=\"pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "Je_fSRQ7O9ss"
   },
   "outputs": [],
   "source": [
    "#@title Convert scores to BigWig (and IGV view)\n",
    "\n",
    "# make bigwigs\n",
    "for tissue in TISSUES[\"description\"].unique():\n",
    "  ranges = scores[[\"chrom\",\"position\",tissue]].copy()\n",
    "  ranges.columns = [\"Chromosome\",\"Start\",\"Score\"]\n",
    "  ranges[\"Chromosome\"] = ranges[\"Chromosome\"]\n",
    "  ranges[\"Start\"] = ranges[\"Start\"] - 1\n",
    "  ranges[\"End\"] = ranges[\"Start\"] + 1\n",
    "  ranges[\"Name\"] = \"U0\"\n",
    "  ranges[\"Score\"] = ranges[\"Score\"]\n",
    "  ranges[\"Strand\"] = strand\n",
    "  ranges = ranges[[\"Chromosome\",\"Start\",\"End\",\"Score\",\"Name\",\"Strand\"]].copy()\n",
    "  ranges = pr.PyRanges(ranges)\n",
    "  chromosome_sizes = {chrom: len(genome[chrom]) for chrom in genome.keys()}\n",
    "  ranges.to_bigwig(f\"{jobname}-{tissue}.bw\", chromosome_sizes, value_col=\"Score\", rpm=False)\n",
    "  # ranges.to_bigwig(dryrun=True, value_col=\"Score\", rpm=False)\n",
    "\n",
    "# make tracks\n",
    "if igv_view:\n",
    "  igv_notebook.init()\n",
    "\n",
    "  if predict_from==\"genomic_coordinates\":\n",
    "    b = igv_notebook.Browser({\n",
    "        \"reference\": {\n",
    "            \"name\": f\"{species}\",\n",
    "            \"fastaPath\": f\"./{fasta_file}\",\n",
    "            \"indexPath\": f\"./{fasta_file}.fai\"\n",
    "        },\n",
    "        \"locus\": f\"{chrom}:{start}-{end}\"\n",
    "    })\n",
    "\n",
    "    b.load_track({\n",
    "          \"name\": \"Gencode v44\",\n",
    "          \"url\": \"https://igv-genepattern-org.s3.amazonaws.com/genomes/hg38/gencode.v44.basic.annotation.gff3.gz\",\n",
    "          \"indexURL\": \"https://igv-genepattern-org.s3.amazonaws.com/genomes/hg38/gencode.v44.basic.annotation.gff3.gz.tbi\",\n",
    "          \"order\": 1,\n",
    "          \"format\": \"gff3\",\n",
    "          \"type\": \"annotation\"\n",
    "    })\n",
    "\n",
    "    b.load_track({\n",
    "          \"name\": \"Refseq Select\",\n",
    "          \"format\": \"refgene\",\n",
    "          \"url\": \"https://hgdownload.soe.ucsc.edu/goldenPath/hg38/database/ncbiRefSeqSelect.txt.gz\",\n",
    "          \"indexed\": False,\n",
    "          \"order\": 2,\n",
    "          \"infoURL\": \"https://www.ncbi.nlm.nih.gov/gene/?term=$$\",\n",
    "          \"type\": \"annotation\",\n",
    "          \"height\": 70\n",
    "    })\n",
    "\n",
    "  else:\n",
    "    b = igv_notebook.Browser({\n",
    "      \"reference\": {\n",
    "          \"name\": \"custom\",\n",
    "          \"fastaPath\": f\"./{fasta_file}\",\n",
    "          \"indexPath\": f\"./{fasta_file}.fai\"\n",
    "      },\n",
    "      \"locus\": f\"{chrom}:{start}-{end}\"\n",
    "    })\n",
    "\n",
    "  colors = sns.color_palette(n_colors=len(TISSUES[\"description\"].unique())).as_hex()[:]\n",
    "  for tissue, color in zip(TISSUES[\"description\"].unique(), colors):\n",
    "    b.load_track({\n",
    "        \"name\": tissue,\n",
    "        \"path\": f\"./{jobname}-{tissue}.bw\",\n",
    "        \"format\": \"bigwig\",\n",
    "        \"type\": \"wig\",\n",
    "        \"color\": color,\n",
    "        \"graphType\": \"line\",\n",
    "    })\n",
    "\n",
    "  b.zoom_in()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 17
    },
    "id": "uKoenWWfmJED",
    "outputId": "452cd94b-0fe9-4820-89fb-54a9f2911e0e"
   },
   "outputs": [],
   "source": [
    "#@title Download data\n",
    "\n",
    "# prep outputs\n",
    "## zip into a folder\n",
    "output_folder = f\"{jobname}-outputs\"\n",
    "os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "shutil.copy(filename_scores, os.path.join(output_folder, filename_scores))\n",
    "shutil.copy(filename_plt, os.path.join(output_folder, filename_plt))\n",
    "for tissue in TISSUES[\"description\"].unique():\n",
    "    shutil.copy(f\"{jobname}-{tissue}.bw\", os.path.join(output_folder, f\"{jobname}-{tissue}.bw\"))\n",
    "\n",
    "shutil.make_archive(output_folder, 'zip', output_folder)\n",
    "\n",
    "## download folder if user wants to\n",
    "if download_outputs:\n",
    "  files.download(f\"{output_folder}.zip\")"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "gpuType": "T4",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
