{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "XSQPhLCv_wVq"
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/MiqG/ColabSplice/blob/main/notebooks/SpliceTransformer.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "-6T7ynmAAjjx"
   },
   "source": [
    "# **SpliceTransformer**\n",
    "\n",
    "## Model information\n",
    "- [Github](https://github.com/ShenLab-Genomics/SpliceTransformer)\n",
    "- [Publication](https://doi.org/10.1038/s41467-024-53088-6)\n",
    "\n",
    "\n",
    "## Tips and Instructions\n",
    "- click the little â–¶ play icon to the left of each cell below. Or Runtime > Run all to give it a try. This will produce a zip file called `test_SMN2.zip`, containing:\n",
    "  - splice site strength predictions:\n",
    "    - table: `test_SMN2-scores.tsv`\n",
    "    - plot: `test_SMN2-splice_site_strength-line.pdf`\n",
    "  - bigwig files that can be visualized through [IGV](https://igv.org/):\n",
    "    - `test_SMN2-Adipose Tissue.bw`\n",
    "    - `test_SMN2-Blood Vessel.bw`\n",
    "    - `test_SMN2-Blood.bw`\n",
    "    - `test_SMN2-Brain.bw`\n",
    "    - `test_SMN2-Colon.bw`\n",
    "    - `test_SMN2-Heart.bw`\n",
    "    - `test_SMN2-Kidney.bw`\n",
    "    - `test_SMN2-Liver.bw`\n",
    "    - `test_SMN2-Lung.bw`\n",
    "    - `test_SMN2-Muscle.bw`\n",
    "    - `test_SMN2-Nerve.bw`\n",
    "    - `test_SMN2-Skin.bw`\n",
    "    - `test_SMN2-Small intestine.bw`\n",
    "    - `test_SMN2-Spleen.bw`\n",
    "    - `test_SMN2-Stomach.bw`\n",
    "     \n",
    "- The notebooks runs perfectly without GPU.\n",
    "\n",
    "## Limitations\n",
    "\n",
    "`SpliceTransformer` only outputs predictions for a maximum of 8192 bp."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "7KnUYg24_Tkp"
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "#@title ## Install packages (~1min 30s)\n",
    "import os, time\n",
    "if not os.path.isfile(\"finished_install\"):\n",
    "  # genomic file handling\n",
    "  os.system(\"pip install -q pyfaidx==0.8.1.1 pyranges pyBigWig pyrle biopython\")\n",
    "\n",
    "  # install SpliceTransformer\n",
    "  os.system(\"git clone https://github.com/ShenLab-Genomics/SpliceTransformer.git\")\n",
    "  os.system(\"mkdir -p SpliceTransformer/model/weights\")\n",
    "  os.system(\"gdown 1d8n4vHDSbXqpPc_JFEswLomSUDBgHvno\")\n",
    "  os.system(\"mv SpTransformer_pytorch.ckpt SpliceTransformer/model/weights/\")\n",
    "  os.system(\"pip install -q pyvcf3 pyensembl sinkhorn-transformer gffutils\")\n",
    "\n",
    "  # install visualization\n",
    "  os.system(\"pip install -q igv-notebook\")\n",
    "\n",
    "  os.system(\"touch finished_install\")\n",
    "\n",
    "# imports\n",
    "import pyfaidx\n",
    "import pyranges as pr\n",
    "import hashlib, re, os\n",
    "import torch\n",
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import igv_notebook\n",
    "import shutil\n",
    "from google.colab import files\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "from torch.nn import functional as F\n",
    "from SpliceTransformer.model.model import SpTransformer\n",
    "\n",
    "TISSUES = ['Adipose Tissue', 'Blood', 'Blood Vessel', 'Brain', 'Colon', 'Heart', 'Kidney',\n",
    "           'Liver', 'Lung', 'Muscle', 'Nerve', 'Small Intestine', 'Skin', 'Spleen', 'Stomach']\n",
    "\n",
    "# notebook-wide variables\n",
    "GENOME_FASTA_URLS = {\n",
    "    \"homo_sapiens (hg38-GencodeV44)\": \"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_44/GRCh38.primary_assembly.genome.fa.gz\"\n",
    "}\n",
    "\n",
    "# functions\n",
    "def install_genome(species, genome_fasta_urls, fasta_file):\n",
    "  if not os.path.isfile(\"finished_install_genome\"):\n",
    "    print(f\"Downloading genome for {species}...\")\n",
    "    os.system(f\"wget {genome_fasta_urls[species]} -O {fasta_file}\")\n",
    "\n",
    "    print(\"Uncompressing genome fasta...\")\n",
    "    os.system(f\"gunzip {fasta_file}\")\n",
    "\n",
    "    os.system(\"touch finished_install_genome\")\n",
    "\n",
    "def get_hash(x):\n",
    "  return hashlib.sha1(x.encode()).hexdigest()\n",
    "\n",
    "def save_fasta(sequence, filename, header=\"sequence\"):\n",
    "  record = SeqRecord(Seq(sequence), id=header, description=\"\")\n",
    "  with open(filename, \"w\") as f:\n",
    "    SeqIO.write(record, f, \"fasta\")\n",
    "\n",
    "def reverse_complement(seq, padval_seq=\"N\"):\n",
    "  complement = str.maketrans(padval_seq+\"ACGTacgt\", padval_seq+\"TGCAtgca\")\n",
    "  return seq.translate(complement)[::-1]\n",
    "\n",
    "def load_model():\n",
    "  weights = torch.load(\"SpliceTransformer/model/weights/SpTransformer_pytorch.ckpt\", weights_only=True, map_location=torch.device('cpu'))\n",
    "  model = SpTransformer(128, context_len=4000, tissue_num=15, max_seq_len=8192, attn_depth=8, training=False)\n",
    "\n",
    "  state_dict = weights[\"state_dict\"].copy()\n",
    "  state_dict[\"attn.pos_emb.weights.0\"] = state_dict[\"attn.pos_emb.weights_0\"]\n",
    "  state_dict[\"attn.pos_emb.weights.1\"] = state_dict[\"attn.pos_emb.weights_1\"]\n",
    "  del state_dict[\"attn.pos_emb.weights_0\"], state_dict[\"attn.pos_emb.weights_1\"]\n",
    "\n",
    "  model.load_state_dict(state_dict)\n",
    "  return model\n",
    "\n",
    "IN_MAP = np.asarray([[0, 0, 0, 0],\n",
    "                     [1, 0, 0, 0],\n",
    "                     [0, 1, 0, 0],\n",
    "                     [0, 0, 1, 0],\n",
    "                     [0, 0, 0, 1]])\n",
    "\n",
    "def one_hot_encode(seq):\n",
    "    '''\n",
    "    Parse input RNA sequence into one-hot-encoding format\n",
    "    '''\n",
    "    seq = seq.upper().replace('A', '1').replace('C', '2')\n",
    "    seq = seq.replace('G', '3').replace('T', '4').replace('N', '0')\n",
    "    seq = np.asarray(list(map(int, list(seq))))\n",
    "    seq = IN_MAP[seq.astype('int8')]\n",
    "    return seq\n",
    "\n",
    "def predict(sequence, strand, model):\n",
    "\n",
    "  print(\"Computing prediction across different tissues...\")\n",
    "\n",
    "  if strand == '-':\n",
    "    sequence = reverse_complement(sequence)\n",
    "\n",
    "  sequence = one_hot_encode(sequence).T\n",
    "  sequence = torch.from_numpy(sequence).float().unsqueeze(0)\n",
    "  if torch.cuda.is_available():\n",
    "    sequence = sequence.cuda()\n",
    "\n",
    "  model.eval()\n",
    "  with torch.no_grad():\n",
    "    # predict\n",
    "    scores = model(sequence).cpu().detach()\n",
    "\n",
    "  # post decorate\n",
    "  splice_site_prob = F.softmax(scores[:, :3, :], dim=1)\n",
    "  scores = torch.sigmoid(scores[:, 3:, :])\n",
    "  scores = scores.numpy()[0,:,:]\n",
    "\n",
    "  if strand == '-':\n",
    "    scores = scores[:,::-1]\n",
    "\n",
    "  scores = pd.DataFrame(scores.T, columns=TISSUES)\n",
    "\n",
    "  return scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "ir1Rl6S2Cfe0"
   },
   "outputs": [],
   "source": [
    "#@title ##run **SpliceTransformer**\n",
    "%%time\n",
    "\n",
    "# user inputs\n",
    "# @markdown Run settings\n",
    "jobname = \"test_SMN2\" #@param {type:\"string\"}\n",
    "jobname = re.sub(r'\\W+', '', jobname)[:50]\n",
    "\n",
    "predict_from = \"genomic_coordinates\" #@param [\"custom_sequence\",\"genomic_coordinates\"]\n",
    "\n",
    "download_outputs = True #@param {type:\"boolean\"}\n",
    "igv_view = True #@param {type:\"boolean\"}\n",
    "\n",
    "# @markdown ---\n",
    "\n",
    "if predict_from == \"genomic_coordinates\":\n",
    "  # @markdown Genomic coordinates (NOTE: requires installing genome at first run (~3min)):\n",
    "  # coordinates\n",
    "  chrom = \"chr5\" #@param {type:\"string\"}\n",
    "  start = 70049669 #@param {type:\"integer\"}\n",
    "  end = 70077595 #@param {type:\"integer\"}\n",
    "  strand = \"+\" #@param [\"-\",\"+\"]\n",
    "  margin = 4000\n",
    "\n",
    "  # genome\n",
    "  species = \"homo_sapiens (hg38-GencodeV44)\" #@param [\"homo_sapiens (hg38-GencodeV44)\"]\n",
    "  fasta_file = os.path.basename(GENOME_FASTA_URLS[species])\n",
    "  install_genome(species, GENOME_FASTA_URLS, fasta_file)\n",
    "  fasta_file = fasta_file.replace(\".gz\",\"\")\n",
    "\n",
    "  print(\"Loading genome data...\")\n",
    "  genome = pyfaidx.Fasta(fasta_file)\n",
    "  sequence = genome[chrom][(start-1-margin):(end+margin)].seq\n",
    "\n",
    "elif predict_from == \"custom_sequence\":\n",
    "  # @markdown ---\n",
    "  # @markdown Custom pre-mRNA sequence (try copy-pasting [FAS sequence](https://www.ncbi.nlm.nih.gov/nuccore/NC_000010.11?report=fasta&from=88964050&to=89017059)):\n",
    "  sequence = \"\" # @param {\"type\":\"string\",\"placeholder\":\"Paste nucleotide sequence here\"}\n",
    "  sequence = sequence.replace(\" \",\"\")\n",
    "  chrom = f\"{jobname}\"\n",
    "  strand = \"+\"\n",
    "  margin = 4000\n",
    "  sequence = margin*\"N\"+sequence+margin*\"N\"\n",
    "\n",
    "  # save custom sequence as fasta\n",
    "  fasta_file = f\"{jobname}.fa\"\n",
    "  save_fasta(sequence[margin:-margin], fasta_file, header=f\"{jobname}\")\n",
    "  genome = pyfaidx.Fasta(fasta_file)\n",
    "  start = 1\n",
    "  end = len(sequence) - margin*2\n",
    "\n",
    "sequence = sequence.upper().replace(\"U\",\"T\")\n",
    "assert set(sequence).issubset({\"A\", \"C\", \"T\", \"G\", \"N\"})\n",
    "\n",
    "ID = jobname+\"_\"+get_hash(sequence)[:5]\n",
    "\n",
    "# load model\n",
    "model = load_model()\n",
    "if torch.cuda.is_available():\n",
    "  model = model.cuda()\n",
    "\n",
    "# make prediction\n",
    "print(f\"Using {predict_from} to make prediction.\")\n",
    "scores = predict(sequence, strand, model)\n",
    "\n",
    "# add more info\n",
    "positions = np.arange(start, end+1)\n",
    "if strand==\"-\":\n",
    "  sequence = reverse_complement(sequence)[::-1]\n",
    "\n",
    "scores[\"sequence\"] = list(sequence[margin:-margin])\n",
    "scores[\"position\"] = positions\n",
    "scores[\"chrom\"] = chrom\n",
    "\n",
    "# save\n",
    "filename_scores = f\"{jobname}-scores.tsv\"\n",
    "scores.to_csv(filename_scores, index=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "OxSVXKsmLGUQ"
   },
   "outputs": [],
   "source": [
    "#@title PDF plot\n",
    "\n",
    "X = scores.melt(id_vars=[\"chrom\",\"position\",\"sequence\"], var_name=\"tissue\")\n",
    "\n",
    "g = sns.relplot(data=X, x=\"position\", y=\"value\", hue=\"tissue\", row=\"tissue\", kind=\"line\", height=1.5, aspect=10)\n",
    "g.set_xlabels(\"\")\n",
    "g.set_ylabels(\"\", clear_inner=False)\n",
    "g.figure.text(0.5, 0.02, \"Nucleotide Position\", ha=\"center\", fontsize=10)\n",
    "g.figure.text(0.0, 0.5, \"Predicted Splice Site Strength\", va=\"center\", rotation=\"vertical\", fontsize=10)\n",
    "\n",
    "filename_plt = f\"{jobname}-splice_site_strength-line.pdf\"\n",
    "g.figure.savefig(filename_plt, format=\"pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "Je_fSRQ7O9ss"
   },
   "outputs": [],
   "source": [
    "#@title Convert scores to BigWig (and IGV view)\n",
    "\n",
    "# make bigwigs\n",
    "for tissue in TISSUES:\n",
    "  ranges = scores[[\"chrom\",\"position\",tissue]].copy()\n",
    "  ranges.columns = [\"Chromosome\",\"Start\",\"Score\"]\n",
    "  ranges[\"Chromosome\"] = ranges[\"Chromosome\"]\n",
    "  ranges[\"Start\"] = ranges[\"Start\"] - 1\n",
    "  ranges[\"End\"] = ranges[\"Start\"] + 1\n",
    "  ranges[\"Name\"] = \"U0\"\n",
    "  ranges[\"Score\"] = ranges[\"Score\"]\n",
    "  ranges[\"Strand\"] = strand\n",
    "  ranges = ranges[[\"Chromosome\",\"Start\",\"End\",\"Score\",\"Name\",\"Strand\"]].copy()\n",
    "  ranges = pr.PyRanges(ranges)\n",
    "  chromosome_sizes = {chrom: len(genome[chrom]) for chrom in genome.keys()}\n",
    "  ranges.to_bigwig(f\"{jobname}-{tissue}.bw\", chromosome_sizes, value_col=\"Score\", rpm=False)\n",
    "  # ranges.to_bigwig(dryrun=True, value_col=\"Score\", rpm=False)\n",
    "\n",
    "# make tracks\n",
    "if igv_view:\n",
    "  igv_notebook.init()\n",
    "\n",
    "  if predict_from==\"genomic_coordinates\":\n",
    "    b = igv_notebook.Browser({\n",
    "        \"reference\": {\n",
    "            \"name\": f\"{species}\",\n",
    "            \"fastaPath\": f\"./{fasta_file}\",\n",
    "            \"indexPath\": f\"./{fasta_file}.fai\"\n",
    "        },\n",
    "        \"locus\": f\"{chrom}:{start}-{end}\"\n",
    "    })\n",
    "\n",
    "    b.load_track({\n",
    "          \"name\": \"Gencode v44\",\n",
    "          \"url\": \"https://igv-genepattern-org.s3.amazonaws.com/genomes/hg38/gencode.v44.basic.annotation.gff3.gz\",\n",
    "          \"indexURL\": \"https://igv-genepattern-org.s3.amazonaws.com/genomes/hg38/gencode.v44.basic.annotation.gff3.gz.tbi\",\n",
    "          \"order\": 1,\n",
    "          \"format\": \"gff3\",\n",
    "          \"type\": \"annotation\"\n",
    "    })\n",
    "\n",
    "    b.load_track({\n",
    "          \"name\": \"Refseq Select\",\n",
    "          \"format\": \"refgene\",\n",
    "          \"url\": \"https://hgdownload.soe.ucsc.edu/goldenPath/hg38/database/ncbiRefSeqSelect.txt.gz\",\n",
    "          \"indexed\": False,\n",
    "          \"order\": 2,\n",
    "          \"infoURL\": \"https://www.ncbi.nlm.nih.gov/gene/?term=$$\",\n",
    "          \"type\": \"annotation\",\n",
    "          \"height\": 70\n",
    "    })\n",
    "\n",
    "  else:\n",
    "    b = igv_notebook.Browser({\n",
    "      \"reference\": {\n",
    "          \"name\": \"custom\",\n",
    "          \"fastaPath\": f\"./{fasta_file}\",\n",
    "          \"indexPath\": f\"./{fasta_file}.fai\"\n",
    "      },\n",
    "      \"locus\": f\"{chrom}:{start}-{end}\"\n",
    "    })\n",
    "\n",
    "  colors = sns.color_palette(n_colors=len(TISSUES)).as_hex()[:]\n",
    "  for tissue, color in zip(TISSUES, colors):\n",
    "    b.load_track({\n",
    "        \"name\": tissue,\n",
    "        \"path\": f\"./{jobname}-{tissue}.bw\",\n",
    "        \"format\": \"bigwig\",\n",
    "        \"type\": \"wig\",\n",
    "        \"color\": color,\n",
    "        \"graphType\": \"line\",\n",
    "    })\n",
    "\n",
    "  b.zoom_in()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 17
    },
    "id": "uKoenWWfmJED",
    "outputId": "71799798-ea81-4174-8e90-4abd15472498"
   },
   "outputs": [],
   "source": [
    "#@title Download data\n",
    "\n",
    "# prep outputs\n",
    "## zip into a folder\n",
    "output_folder = f\"{jobname}-outputs\"\n",
    "os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "shutil.copy(filename_scores, os.path.join(output_folder, filename_scores))\n",
    "shutil.copy(filename_plt, os.path.join(output_folder, filename_plt))\n",
    "for tissue in TISSUES:\n",
    "    shutil.copy(f\"{jobname}-{tissue}.bw\", os.path.join(output_folder, f\"{jobname}-{tissue}.bw\"))\n",
    "\n",
    "shutil.make_archive(output_folder, 'zip', output_folder)\n",
    "\n",
    "## download folder if user wants to\n",
    "if download_outputs:\n",
    "  files.download(f\"{output_folder}.zip\")"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
