{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "XSQPhLCv_wVq"
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/MiqG/ColabSplice/blob/main/notebooks/SpliceAI.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "-6T7ynmAAjjx"
   },
   "source": [
    "# **SpliceAI**\n",
    "\n",
    "## Model information\n",
    "- [Github](https://github.com/Illumina/SpliceAI)\n",
    "- [Publication](https://doi.org/10.1016/j.cell.2018.12.015)\n",
    "\n",
    "\n",
    "## Tips and Instructions\n",
    "- click the little â–¶ play icon to the left of each cell below. Or Runtime > Run all to give it a try. This will produce a zip file called `test_SMN2.zip`, containing:\n",
    "  - splice site strength predictions:\n",
    "    - table: `test_SMN2-scores.tsv`\n",
    "    - plot: `test_SMN2-splice_site_strength-line.pdf`\n",
    "  - bigwig files that can be visualized through [IGV](https://igv.org/):\n",
    "    - `test_SMN2-acceptor.bw`\n",
    "    - `test_SMN2-donor.bw`\n",
    "     \n",
    "- The notebooks runs perfectly without GPU.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "7KnUYg24_Tkp"
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "#@title ## Install packages (~1min 30s)\n",
    "import os, time\n",
    "if not os.path.isfile(\"finished_install\"):\n",
    "  # genomic file handling\n",
    "  os.system(\"pip install -q pyfaidx==0.8.1.1 pyranges pyBigWig pyrle biopython\")\n",
    "\n",
    "  # install SpliceAI\n",
    "  os.system(\"pip install -q spliceai\")\n",
    "\n",
    "  # install visualization\n",
    "  os.system(\"pip install -q igv-notebook\")\n",
    "\n",
    "  os.system(\"touch finished_install\")\n",
    "\n",
    "# imports\n",
    "import pyfaidx\n",
    "import pyranges as pr\n",
    "import hashlib, re, os\n",
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import igv_notebook\n",
    "import shutil\n",
    "from google.colab import files\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "from keras.models import load_model\n",
    "from pkg_resources import resource_filename\n",
    "from spliceai.utils import one_hot_encode\n",
    "import gc\n",
    "\n",
    "# notebook-wide variables\n",
    "GENOME_FASTA_URLS = {\n",
    "    \"homo_sapiens (hg38-GencodeV44)\": \"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_44/GRCh38.primary_assembly.genome.fa.gz\"\n",
    "}\n",
    "# genome_gff_urls = {\n",
    "#     \"homo_sapiens (hg38-Ensembl113)\": \"https://ftp.ensembl.org/pub/release-113/gff3/homo_sapiens/Homo_sapiens.GRCh38.113.gff3.gz\"\n",
    "# }\n",
    "\n",
    "# functions\n",
    "def install_genome(species, genome_fasta_urls, fasta_file):\n",
    "  if not os.path.isfile(\"finished_install_genome\"):\n",
    "    print(f\"Downloading genome for {species}...\")\n",
    "    os.system(f\"wget {genome_fasta_urls[species]} -O {fasta_file}\")\n",
    "\n",
    "    print(\"Uncompressing genome fasta...\")\n",
    "    os.system(f\"gunzip {fasta_file}\")\n",
    "\n",
    "    os.system(\"touch finished_install_genome\")\n",
    "\n",
    "def get_hash(x):\n",
    "  return hashlib.sha1(x.encode()).hexdigest()\n",
    "\n",
    "def save_fasta(sequence, filename, header=\"sequence\"):\n",
    "  record = SeqRecord(Seq(sequence), id=header, description=\"\")\n",
    "  with open(filename, \"w\") as f:\n",
    "    SeqIO.write(record, f, \"fasta\")\n",
    "\n",
    "def reverse_complement(seq, padval_seq=\"N\"):\n",
    "  complement = str.maketrans(padval_seq+\"ACGTacgt\", padval_seq+\"TGCAtgca\")\n",
    "  return seq.translate(complement)[::-1]\n",
    "\n",
    "def load_models():\n",
    "  paths = ('models/spliceai{}.h5'.format(x) for x in range(1, 6))\n",
    "  models = [load_model(resource_filename('spliceai', x)) for x in paths]\n",
    "  return models\n",
    "\n",
    "def predict(sequence, strand, models):\n",
    "  print(\"Computing predictions...\")\n",
    "\n",
    "  if strand == \"-\":\n",
    "    sequence = reverse_complement(sequence)\n",
    "\n",
    "  x = one_hot_encode(sequence)[None, :]\n",
    "  y = np.mean([models[m].predict(x) for m in range(5)], axis=0)\n",
    "  acceptor_prob = y[0, :, 1]\n",
    "  donor_prob = y[0, :, 2]\n",
    "  scores = {\"acceptor\": acceptor_prob, \"donor\": donor_prob}\n",
    "\n",
    "  if strand == \"-\":\n",
    "    scores = {k: v[::-1] for k, v in scores.items()}\n",
    "\n",
    "  scores = pd.DataFrame(scores)\n",
    "\n",
    "  return scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "ir1Rl6S2Cfe0"
   },
   "outputs": [],
   "source": [
    "#@title ##run **SpliceAI**\n",
    "%%time\n",
    "\n",
    "# user inputs\n",
    "# @markdown Run settings\n",
    "jobname = \"test_SMN2\" #@param {type:\"string\"}\n",
    "jobname = re.sub(r'\\W+', '', jobname)[:50]\n",
    "\n",
    "predict_from = \"genomic_coordinates\" #@param [\"custom_sequence\",\"genomic_coordinates\"]\n",
    "\n",
    "download_outputs = True #@param {type:\"boolean\"}\n",
    "igv_view = True #@param {type:\"boolean\"}\n",
    "\n",
    "# @markdown ---\n",
    "\n",
    "if predict_from == \"genomic_coordinates\":\n",
    "  # @markdown Genomic coordinates (NOTE: requires installing genome at first run (~3min)):\n",
    "  # coordinates\n",
    "  chrom = \"chr5\" #@param {type:\"string\"}\n",
    "  start = 70049669 #@param {type:\"integer\"}\n",
    "  end = 70077595 #@param {type:\"integer\"}\n",
    "  strand = \"+\" #@param [\"-\",\"+\"]\n",
    "  margin = 5000\n",
    "\n",
    "  # genome\n",
    "  species = \"homo_sapiens (hg38-GencodeV44)\" #@param [\"homo_sapiens (hg38-GencodeV44)\"]\n",
    "  fasta_file = os.path.basename(GENOME_FASTA_URLS[species])\n",
    "  install_genome(species, GENOME_FASTA_URLS, fasta_file)\n",
    "  fasta_file = fasta_file.replace(\".gz\",\"\")\n",
    "\n",
    "  print(\"Loading genome data...\")\n",
    "  genome = pyfaidx.Fasta(fasta_file)\n",
    "  # gtf = pr.read_gtf(gtf_file)\n",
    "  sequence = genome[chrom][(start-1-margin):(end+margin)].seq\n",
    "\n",
    "elif predict_from == \"custom_sequence\":\n",
    "  # @markdown ---\n",
    "  # @markdown Custom pre-mRNA sequence (try copy-pasting [FAS sequence](https://www.ncbi.nlm.nih.gov/nuccore/NC_000010.11?report=fasta&from=88964050&to=89017059)):\n",
    "  sequence = \"\" # @param {\"type\":\"string\",\"placeholder\":\"Paste nucleotide sequence here\"}\n",
    "  sequence = sequence.replace(\" \",\"\")\n",
    "  chrom = f\"{jobname}\"\n",
    "  strand = \"+\"\n",
    "  margin = 5000\n",
    "  sequence = margin*\"N\"+sequence+margin*\"N\"\n",
    "\n",
    "  # save custom sequence as fasta\n",
    "  fasta_file = f\"{jobname}.fa\"\n",
    "  save_fasta(sequence[margin:-margin], fasta_file, header=f\"{jobname}\")\n",
    "  genome = pyfaidx.Fasta(fasta_file)\n",
    "  start = 1\n",
    "  end = len(sequence) - margin*2\n",
    "\n",
    "sequence = sequence.upper().replace(\"U\",\"T\")\n",
    "assert set(sequence).issubset({\"A\", \"C\", \"T\", \"G\", \"N\"})\n",
    "\n",
    "length = len(sequence)\n",
    "print(\"length\",length)\n",
    "\n",
    "ID = jobname+\"_\"+get_hash(sequence)[:5]\n",
    "\n",
    "# prep model(s)\n",
    "if \"models\" not in dir():\n",
    "  if \"models\" in dir():\n",
    "    # delete old model from memory\n",
    "    del models\n",
    "    gc.collect()\n",
    "\n",
    "models = load_models()\n",
    "\n",
    "# make prediction\n",
    "print(f\"Using {predict_from} to make prediction.\")\n",
    "scores = predict(sequence, strand, models)\n",
    "\n",
    "# add more info\n",
    "positions = np.arange(start, end+1)\n",
    "if strand==\"-\":\n",
    "  sequence = reverse_complement(sequence)[::-1]\n",
    "\n",
    "scores[\"sequence\"] = list(sequence[margin:-margin])\n",
    "scores[\"position\"] = positions\n",
    "scores[\"chrom\"] = chrom\n",
    "\n",
    "# save\n",
    "filename_scores = f\"{jobname}-scores.tsv\"\n",
    "scores.to_csv(filename_scores, index=False, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "OxSVXKsmLGUQ"
   },
   "outputs": [],
   "source": [
    "#@title PDF plot\n",
    "\n",
    "X = scores.melt(id_vars=[\"chrom\",\"position\",\"sequence\"], var_name=\"splice_site\")\n",
    "\n",
    "g = sns.relplot(data=X, x=\"position\", y=\"value\", hue=\"splice_site\", row=\"splice_site\", kind=\"line\", height=1.5, aspect=10)\n",
    "g.set_xlabels(\"\")\n",
    "g.set_ylabels(\"\", clear_inner=False)\n",
    "g.figure.text(0.5, 0.02, \"Nucleotide Position\", ha=\"center\", fontsize=10)\n",
    "g.figure.text(0.0, 0.5, \"Predicted Splice Site Strength\", va=\"center\", rotation=\"vertical\", fontsize=10)\n",
    "\n",
    "filename_plt = f\"{jobname}-splice_site_strength-line.pdf\"\n",
    "g.figure.savefig(filename_plt, format=\"pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "Je_fSRQ7O9ss"
   },
   "outputs": [],
   "source": [
    "#@title Convert scores to BigWig (and IGV view)\n",
    "\n",
    "# make bigwigs\n",
    "splice_sites = [\"donor\",\"acceptor\"]\n",
    "for splice_site in splice_sites:\n",
    "  ranges = scores[[\"chrom\",\"position\",splice_site]].copy()\n",
    "  ranges.columns = [\"Chromosome\",\"Start\",\"Score\"]\n",
    "  ranges[\"Chromosome\"] = ranges[\"Chromosome\"]\n",
    "  ranges[\"Start\"] = ranges[\"Start\"] - 1\n",
    "  ranges[\"End\"] = ranges[\"Start\"] + 1\n",
    "  ranges[\"Name\"] = \"U0\"\n",
    "  ranges[\"Score\"] = ranges[\"Score\"]#.round(2)\n",
    "  ranges[\"Strand\"] = strand\n",
    "  ranges = ranges[[\"Chromosome\",\"Start\",\"End\",\"Score\",\"Name\",\"Strand\"]].copy()\n",
    "  ranges = pr.PyRanges(ranges)\n",
    "  chromosome_sizes = {chrom: len(genome[chrom]) for chrom in genome.keys()}\n",
    "  ranges.to_bigwig(f\"{jobname}-{splice_site}.bw\", chromosome_sizes, value_col=\"Score\", rpm=False)\n",
    "  # ranges.to_bigwig(dryrun=True, value_col=\"Score\", rpm=False)\n",
    "\n",
    "# make tracks\n",
    "if igv_view:\n",
    "  igv_notebook.init()\n",
    "\n",
    "  if predict_from==\"genomic_coordinates\":\n",
    "    b = igv_notebook.Browser({\n",
    "        \"reference\": {\n",
    "            \"name\": f\"{species}\",\n",
    "            \"fastaPath\": f\"./{fasta_file}\",\n",
    "            \"indexPath\": f\"./{fasta_file}.fai\"\n",
    "        },\n",
    "        \"locus\": f\"{chrom}:{start}-{end}\"\n",
    "    })\n",
    "\n",
    "    b.load_track({\n",
    "          \"name\": \"Gencode v44\",\n",
    "          \"url\": \"https://igv-genepattern-org.s3.amazonaws.com/genomes/hg38/gencode.v44.basic.annotation.gff3.gz\",\n",
    "          \"indexURL\": \"https://igv-genepattern-org.s3.amazonaws.com/genomes/hg38/gencode.v44.basic.annotation.gff3.gz.tbi\",\n",
    "          \"order\": 1,\n",
    "          \"format\": \"gff3\",\n",
    "          \"type\": \"annotation\"\n",
    "    })\n",
    "\n",
    "    b.load_track({\n",
    "          \"name\": \"Refseq Select\",\n",
    "          \"format\": \"refgene\",\n",
    "          \"url\": \"https://hgdownload.soe.ucsc.edu/goldenPath/hg38/database/ncbiRefSeqSelect.txt.gz\",\n",
    "          \"indexed\": False,\n",
    "          \"order\": 2,\n",
    "          \"infoURL\": \"https://www.ncbi.nlm.nih.gov/gene/?term=$$\",\n",
    "          \"type\": \"annotation\",\n",
    "          \"height\": 70\n",
    "    })\n",
    "\n",
    "  else:\n",
    "    b = igv_notebook.Browser({\n",
    "      \"reference\": {\n",
    "          \"name\": \"custom\",\n",
    "          \"fastaPath\": f\"./{fasta_file}\",\n",
    "          \"indexPath\": f\"./{fasta_file}.fai\"\n",
    "      },\n",
    "      \"locus\": f\"{chrom}:{start}-{end}\"\n",
    "    })\n",
    "\n",
    "  colors = [\"blue\", \"orange\", \"green\", \"red\"]\n",
    "  for splice_site, color in zip(splice_sites, colors):\n",
    "    b.load_track({\n",
    "        \"name\": splice_site,\n",
    "        \"path\": f\"./{jobname}-{splice_site}.bw\",\n",
    "        \"format\": \"bigwig\",\n",
    "        \"type\": \"wig\",\n",
    "        \"color\": color,\n",
    "        \"graphType\": \"line\",\n",
    "    })\n",
    "\n",
    "  b.zoom_in()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "uKoenWWfmJED"
   },
   "outputs": [],
   "source": [
    "#@title Download data\n",
    "\n",
    "# prep outputs\n",
    "## zip into a folder\n",
    "output_folder = f\"{jobname}-outputs\"\n",
    "os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "shutil.copy(filename_scores, os.path.join(output_folder, filename_scores))\n",
    "shutil.copy(filename_plt, os.path.join(output_folder, filename_plt))\n",
    "for splice_site in splice_sites:\n",
    "    shutil.copy(f\"{jobname}-{splice_site}.bw\", os.path.join(output_folder, f\"{jobname}-{splice_site}.bw\"))\n",
    "\n",
    "shutil.make_archive(output_folder, 'zip', output_folder)\n",
    "\n",
    "## download folder if user wants to\n",
    "if download_outputs:\n",
    "  files.download(f\"{output_folder}.zip\")"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
